File: java-algo/src/main/java/com/leetcode/Problem1125.java
package com.leetcode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 1125. Smallest Sufficient Team
 * https://leetcode.cn/problems/smallest-sufficient-team/
 */
public class Problem1125 {
    public int[] smallestSufficientTeam(String[] req_skills, List<List<String>> people) {
        int n = req_skills.length, m = people.size();
        Map<String, Integer> skillIndex = new HashMap<>();
        for (int i = 0; i < n; i++) {
            skillIndex.put(req_skills[i], i);
        }
        List<Integer>[] dp = new List[1 << n];
        dp[0] = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            int curSkill = 0;
            for (String s : people.get(i)) {
                curSkill |= 1 << skillIndex.get(s);
            }
            for (int prev = 0; prev < dp.length; prev++) {
                if (dp[prev] == null) {
                    continue;
                }
                int comb = prev | curSkill;
                if (dp[comb] == null || dp[prev].size() + 1 < dp[comb].size()) {
                    dp[comb] = new ArrayList<>(dp[prev]);
                    dp[comb].add(i);
                }
            }
        }
        return dp[(1 << n) - 1].stream().mapToInt(i -> i).toArray();
    }
}

File: python-algo/leetcode/problem_1125.py
"""1125. Smallest Sufficient Team
https://leetcode.com/problems/smallest-sufficient-team/
"""
from typing import List


class Solution:
    def smallestSufficientTeam(self, req_skills: List[str], people: List[List[str]]) -> List[int]:
        m, n = len(people), len(req_skills)
        skill_index = dict()
        for i, skill in enumerate(req_skills):
            skill_index[i] = skill
        dp = [[] for _ in range(1 << n)]
        for i in range(m):
            cur_skill = 0
            for s in people[i]:
                cur_skill |= 1 << skill_index[s]
            for j in range(len(dp)):
                if not dp[j]:
                    continue
                cmb = j | cur_skill
                if not dp[cmb] or len(dp[j]) + 1 < len(dp[cmb]):
                    dp[cmb] = dp[j][:]
                    dp[cmb].append(i)
        return dp[-1]
